File: mojang-rules/lib/features.ts
import { z } from 'zod';

export type FeatureConstraint = Record<string, boolean>;

export const FeatureConstraintSchema = z.record(z.string(), z.boolean());

export function satisfiesFeatures(
  constraint: FeatureConstraint,
  feats: string[],
): boolean {
  return Object.entries(constraint).every(
    ([feature, should]) => feats.some((f) => f === feature) === should,
  );
}

File: crates/opys-mojang-rules/src/features.rs
use std::collections::BTreeMap;

/// Feature constraint: `{ feature_name: required_value }`. A feature is
/// "satisfied" iff its presence in the active set matches the required bool.
pub type FeatureConstraint = BTreeMap<String, bool>;

pub fn satisfies_features(constraint: &FeatureConstraint, feats: &[String]) -> bool {
    constraint
        .iter()
        .all(|(feature, &should)| feats.iter().any(|f| f == feature) == should)
}
